### DISCUSSION TITLE
Automate frequent set of local git commands

### DISCUSSION THREAD

**Discussion by smontanaro** [community/community#22010] [New to GitHub]
<p>I’ve never been a super sophisticated git user. I am working on a personal website. I have main and dev branches. Unless I’m working on significant functional changes (where i create a specific development branch for that work), my normal pattern is</p>
<p>On my (Linux) laptop:</p>
<ul>
<li>implement changes on dev</li>
<li>run test suite</li>
<li>lather rinse repeat until satisfied</li>
<li>commit &amp; push</li>
<li>check out main</li>
<li>merge from dev &amp; push</li>
</ul>
<p>On my (Linux) server:</p>
<ul>
<li>pull (always on main there)</li>
<li>restart gunicorn (and every now and then nginx, depending on the changes)</li>
</ul>
<p>This works fine, except you’ll notice that I failed to checkout dev again on my laptop, so if I don’t remember to do that, the next time I go to do some work, I will begin that work on main. Most of the time I realize my mistake, then stash, checkout, stash pop. Sometimes I don’t remember until I’ve committed something and correcting things is a bit more involved.</p>
<p>Ignoring the fact that I should deploy my website as a container (I will eventually get there), is there a way within git to automate the commit/push/checkout/merge/push workflow so that I always wind up back on my dev branch, short of just writing a shell script to force this sequence of commands? There’s nothing wrong with the shell script solution. I’m just hoping to improve my git-fu.</p>

**Comment by mike-kaminski** [comment]
<p>You could set up a bash alias/function that does the whole loop for you. I have one, for example, that gets the list of repos in an organization from the API and then clones them all to the directory of my choice. And others for various semi-routine patterns that I run, cleaning out old/closed branches, clearing out leftover tags after promoting an RC to a RELEASE, etc.</p>
<p>You could also just write a bash script and chuck it in /usr/local/bin, same idea for the most part but you don’t run the risk of inadvertently breaking your shell.</p>

**Comment by smontanaro** [comment]
<p>Thanks. I’m sure I can write a shell function or script to do this. I was hoping there was a way to do this completely within <code>git</code>, maybe by tweaking the configuration in my repo. Again, hoping to improve my git-fu. My shell-fu doesn’t really need work. :wink:</p>

**Comment by mike-kaminski** [comment, marked as answer]
<p>Perhaps you could make use of <a href="https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks" rel="noopener nofollow ugc">Git Hooks (git-scm.com)</a>?</p>

**Comment by smontanaro** [comment]
<p>Yes, thanks. That looks like it will work. (Again, I’m not much of a git expert, so this never occurred to me. In retrospect, it probably should have. I’m well-acquainted with hooks in Emacs - the mother of all hookable applications - for instance. Not realizing git has such a feature sorta makes no sense.</p>
